src/JBController.sol

```diff
@@ -803,46 +808,92 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
                 if (split.hook != IJBSplitHook(address(0))) {
-                    // Get a reference to the project's token.
+                    // Mint the tokens for the split hook.
+                    TOKENS.mintFor(address(split.hook), projectId, splitAmount);
+
+                    // Get a reference to the project's token. This will return the 0 address if the project doesn't yet
+                    // have a token.
                     IJBToken token = TOKENS.tokenOf(projectId);
 
                     // Process.
                     split.hook.processSplitWith(
                         JBSplitHookContext({
                             token: address(token),
-                            amount: tokenCount,
-                            decimals: token.decimals(),
+                            amount: splitAmount,
+                            decimals: 18, // Hardcoded in JBTokens.
                             projectId: projectId,
                             groupId: groupId,
                             split: split
                         })
                     );
+                    // If there's a project ID, try to pay the project. If it fails, fallback to paying the beneficiary.
+                } else {
+                    // Use the split's beneficiary if provided, otherwise use the msg sender.
+                    address beneficiary = split.beneficiary != address(0) ? split.beneficiary : _msgSender();
+
+                    if (split.projectId != 0) {
+                        // Get a reference to the project's token. This will return the 0 address if the project doesn't
+                        // yet
+                        // have a token.
+                        IJBToken token = TOKENS.tokenOf(projectId);
+
+                        // Get a reference to the project's payment terminal that accepts the token.
+                        IJBTerminal terminal = token == IJBToken(address(0))
+                            ? IJBTerminal(address(0))
+                            : DIRECTORY.primaryTerminalOf(split.projectId, address(token));
+
+                        // If the paying project doesn't have a token or the receiving project isn't accepting the
+                        // token,
+                        // fallback to sending to the beneficiary.
+                        if (address(token) == address(0) || address(terminal) == address(0)) {
+                            // Mint the tokens.
+                            TOKENS.mintFor(beneficiary, projectId, splitAmount);
+                        } else {
+                            // Mint the tokens to this contract.
+                            TOKENS.mintFor(address(0), projectId, splitAmount);
+
+                            // Send the projectId in the metadata.
+                            bytes memory metadata = bytes(abi.encodePacked(projectId));
+
+                            // Try to fulfill the payment
```

Comment:
shouldn't we `TOKENS.approve(terminal, splitAmount)` here?

Reply:
yes!